Avoid implicitly converting NodesType to ValueType (#420)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

| `Nodes(nl)`       | A list of zero or more nodes, e.g., from a `filter-path` resulting in the nodelist `nl`, which may or may not be empty  |
{: #tbl-typerep title="Concrete representations of abstract instances"}

### Type Conversion {#type-conv}

The following implicit type conversion may occur:

* Where a member of `NodesType` needs to be converted to a
  `LogicalType`, the conversion proceeds as follows:
  * If the nodelist contains one or more nodes, the conversion result
    is `LogicalTrue`.
  * If the nodelist is empty, the conversion result is `LogicalFalse`.

> An expression with a declared type of `NodesType` can be converted to a `ValueType` using the explicit type conversion function `value()` (see {{value}}).

Reviewer: I'm not even sure this is a "conversion" function (there may be several semantics for such a conversion function, and the merit of the present PR is that it makes that function explicit so others can be added later).

Author: Yes, let's avoid that terminology. Adopted your suggestion.

> An expression with a declared type of `NodesType` can be converted to a `ValueType` using the explicit type conversion function `value()` (see {{value}}).

Reviewer: ```suggestion
Given an expression with a declared type of `NodesType`, a `ValueType` can be obtained using a function such as `value()` (see {{value}}).
```

substring exists and `LogicalFalse` otherwise.

### `value` Function Extension {#value}

Parameters:
: 1. `NodesType`

Result:
: `ValueType`

The "value" function extension provides a way to convert a instance of `NodesType` to a value and
make that available for further processing in the filter expression:

~~~ JSONPath
$[?value(@..color) == "red"]
~~~

Its only argument is an instance of `NodesType` (possibly taken from a
`filter-path` as in the example above).  The result is an
instance of `ValueType`.

* If the argument contains a single node, the result is
  the value of the node.
* If the argument is `Nothing` or contains multiple nodes, the

>   result is `Nothing`.

Reviewer 2: This is a gotcha.  It produces potentially unexpected behavior in expressions.

(Yes, I understand that its behavior is specified, but users won't read the spec.)

This is no different than the implicit conversion that exists in the current `main` document.  The problem is the surprise.  This is why I included the table of behaviors that has been deleted in this PR.

The behavior we define shouldn't contain surprises for users.  That just generates StackOverflow questions.  It should be preferred to reduce the number of questions by devising a system that is intuitive.  This isn't it.  (Arguably, what's in `main` also isn't it.)

Author: The point of putting this behaviour inside a function is to make opting in to the behaviour a conscious choice.

> ### Examples

Reviewer 2: We should include examples.

Author: Yes: added.
